### ISSUE TITLE
Wrong compilation error in type combining intersection of unions

### ISSUE THREAD

**Issue by arnauorriols** [microsoft/TypeScript#31556]
**TypeScript Version:**  typescript@3.5.0-dev.20190523

**Search Terms:** "intersection of unions"

**Code**

```ts
type Foo = (A | B) & (C | D);

interface A {
	a: true;
}

interface B {
	a: false;
}

interface C {
	b: "foo";
}

interface D {
	b: "bar";
}

const foo: Foo = {
	a: true,
	// b: "foo"
};
```

**Expected behavior:** Compilation error:
> Property 'b' is missing in type '{ name: string; a: true}' but required in type 'C'.

**Actual behavior:** Compilation error:
> test.ts(19,7): error TS2322: Type '{ a: true; }' is not assignable to type 'Foo'.
  Type '{ a: true; }' is not assignable to type 'B & D'.
    Type '{ a: true; }' is not assignable to type 'B'.
      Types of property 'a' are incompatible.
        Type 'true' is not assignable to type 'false'.

If `// b: "foo"` line is uncommented, code compiles, as expected. 

**Playground Link:** [link](https://www.typescriptlang.org/play/#src=type%20Foo%20%3D%20%7B%0D%0A%09name%3A%20string%3B%0D%0A%7D%20%26%20(A%20%7C%20B)%20%26%20(C%20%7C%20D)%3B%0D%0A%0D%0Ainterface%20A%20%7B%0D%0A%09a%3A%20true%3B%0D%0A%7D%0D%0A%0D%0Ainterface%20B%20%7B%0D%0A%09a%3A%20false%3B%0D%0A%7D%0D%0A%0D%0Ainterface%20C%20%7B%0D%0A%09b%3A%20%22foo%22%3B%0D%0A%7D%0D%0A%0D%0Ainterface%20D%20%7B%0D%0A%09b%3A%20%22bar%22%3B%0D%0A%7D%0D%0A%0D%0Aconst%20foo%3A%20Foo%20%3D%20%7B%0D%0A%20%20%20%20name%3A%20%22foo%22%2C%0D%0A%20%20%20%20a%3A%20true%2C%0D%0A%09%2F%2F%20b%3A%20%22foo%22%0D%0A%7D%3B)

**Related Issues:** 
- https://github.com/microsoft/TypeScript/issues/8896

**Comment by RyanCavanaugh** [comment]
Works as expected now
